src/v/cluster/partition_balancer_planner.cc

```diff
@@ -0,0 +1,309 @@
+        size_t index = 0;
+        while (float(size_to_remove + node_space.free_space)
+                   / float(node_space.total_space)
+                 > _config.max_disc_usage
+               && index < ntp_on_node_sizes.size()) {
+            const auto& partition_to_move = ntp_on_node_sizes[index].second;
+            const auto& topic_metadata = _topic_table.local().topics_map().at(
+              model::topic_namespace(
+                partition_to_move.ns, partition_to_move.tp.topic));
+            std::vector<model::broker_shard> stable_replicas;
+            for (const auto& r : topic_metadata.get_assignments()
+                                   .find(partition_to_move.tp.partition)
+                                   ->replicas) {
+                if (
+                  std::find_if(
+                    nodes_with_disk_constaints_violation.begin(),
+                    nodes_with_disk_constaints_violation.end(),
+                    [&r](const NodeSpace& ns) {
+                        return ns.node_id == r.node_id;
+                    })
+                  == nodes_with_disk_constaints_violation.end()) {
+                    stable_replicas.push_back(
+                      model::broker_shard{r.node_id, r.shard});
+                }
+            }
+            auto new_allocation_units = get_reallocation(
+              *topic_metadata.get_assignments().find(
+                partition_to_move.tp.partition),
+              topic_metadata,
+              stable_replicas,
+              health_report,
+              ntp_on_node_sizes[index].first);
+            if (
+              !new_allocation_units
+              || new_allocation_units.value().get_assignments().size() == 0) {
+                vlog(
+                  clusterlog.info,
+                  "Can't reallocate {} {} {} from down node",
+                  partition_to_move.ns,
+                  partition_to_move.tp.topic,
+                  partition_to_move.tp.partition);
+            } else {
+                size_to_remove += ntp_on_node_sizes[index].first;
+                reassigments.emplace_back(ntp_reassigments{
+                  .ntp = model::ntp(
+                    partition_to_move.ns,
+                    partition_to_move.tp.topic,
+                    partition_to_move.tp.partition),
+                  .allocation_units = std::move(new_allocation_units.value())});
+                _reallocation_ordered += 1;
+            }
+            index++;
+            if (_reallocation_ordered == _config.batch_size) {
+                return;
+            }
+        }
+    }
+}
+
+std::vector<ntp_reassigments> partition_balancer_planner::get_ntp_reassigments(
+  cluster_health_report& health_report) {
+    std::vector<ntp_reassigments> reassigments;
+    // Reset reallocations amount for current cycle
+    _reallocation_ordered = 0;
+
+    // Collect unavailable nodes
+    for (const auto& n : health_report.node_states) {
+        if (!n.is_alive) {
```

Comment:
we should use node_availability_timeout_sec to determine node unavailability

Reply:
fixed